### ISSUE TITLE
Output is Gibberish

### ISSUE THREAD

**Issue by choprahetarth** [Tandemn-Labs/tandemn-vllm#15]
Okay - just tested this. The speed is fine, but the output is completely messed up. This is NOT IDEAL. DIg into it.

**Comment by choprahetarth** [comment]
Even during a single forward pass (loading ALL the model layers on a single peer), the output is gibberish. Need to fix this. 
Likely problems are - 
1) Model weights are not being loaded correctly (dummy=True). 
2) Dtype Conversion between hooks
3) Tensor Shape Conversion 
4) Incorrectly sending the sampler states

Need to check all of them one by one

**Comment by choprahetarth** [comment]
Checked the second. DType conversion is not an issue. However, it might be in case if the hidden states are bfloat16 and are converted to numpy and are sent. Edge case, will fix it soon. However, right now everything is in fp16.

**Comment by choprahetarth** [comment]
Worked at the end - tested it only for non-quantised models. The reason was that the actual lm_head (the layer responsible for producing logits) was being overwritten by model_weights=dummy. I encountered this error a long time ago; I assumed I had fixed it.

**Comment by choprahetarth** [comment]
Issue not fully solved yet - many times different model architectures have tied weights (embeddings and lm_head) being the same. Adding a feature to check if lm_head weights are not found- just copy the stuff from embeddings over to lm_head as a direct replacement. 

This solves the problem. Need to fully make it modular and check the code quality for the same.
